clustermesh: correctly remove remoteCache on connection disruption

squeed opened the pull request:
Upon reconnect, we failed to remove the old remoteCache (we were looking at the wrong Allocator on cleanup), meaning that every time we reconnected, all old remoteCaches were kept around.

This was, at best, a memory leak, and at worst meant that we continued to read stale data even after reconnecting, depending on the ordering of a map iteration.

Many thanks to @oblazek, who found the root cause.

Fixes: #22988
Fixes: #13446

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

```release-note
Fixes a memory leak and (possible) source of stale data for Clustermesh whenever the connection to the remote cluster is disrupted or restarted.
```

squeed committed:
clustermesh: correctly remove remoteCache on connection disruption

Upon reconnect, we failed to remove the old remoteCache (we were looking
at the wrong Allocator on cleanup), meaning that every time we
reconnected, all old remoteCaches were kept around.

This was, at best, a memory leak, and at worst meant that we continued
to read stale data even after reconnecting, depending on the ordering of
a map iteration.

Fixes: #22988
Fixes: #13446

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

christarazi reviewed (commented):
Thanks for submitting the fix and thanks to @oblazek for finding it!

christarazi reviewed (approved):
Thanks for submitting the fix and thanks to @oblazek for finding it!

aditighag commented on pkg/identity/cache/allocator.go:
```diff
@@ -440,18 +440,24 @@ func (m *CachingIdentityAllocator) ReleaseSlice(ctx context.Context, owner Ident
 // WatchRemoteIdentities starts watching for identities in another kvstore and
 // syncs all identities to the local identity cache.
-func (m *CachingIdentityAllocator) WatchRemoteIdentities(backend kvstore.BackendOperations) (*allocator.RemoteCache, error) {
+func (m *CachingIdentityAllocator) WatchRemoteIdentities(name string, backend kvstore.BackendOperations) (*allocator.RemoteCache, error) {
```
Nit: Please update the function description.

aditighag reviewed (approved):
Good find! The refactoring/fixing typos related changes could've been separated in a separate commit. LGTM. 
Some of the functions need updated description.

oblazek commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -269,6 +269,7 @@ func (rc *remoteCluster) restartRemoteConnection(allocator RemoteIdentityWatcher
 				rc.releaseOldConnection()
 				rc.mesh.metricTotalNodes.WithLabelValues(rc.mesh.conf.Name, rc.mesh.conf.NodeName, rc.name).Set(float64(rc.remoteNodes.NumEntries()))
 				rc.mesh.metricReadinessStatus.WithLabelValues(rc.mesh.conf.Name, rc.mesh.conf.NodeName, rc.name).Set(metrics.BoolToFloat64(rc.isReadyLocked()))
+				allocator.RemoveRemoteIdentities(rc.name)
```
can we add this also somewhere [here](https://github.com/cilium/cilium/blob/master/pkg/clustermesh/remote_cluster.go#L165)? otherwise it does not really fix the memleak issue.. the controller itself is not stopped, so this cleanup is never run in my case...

squeed replied:
@oblazek There should be only one controller per remote cluster, so as soon as a new connection is made, it will replace the old cache in the allocator. I decided to preserve the old one, so that connections aren't disrupted while etcd blips. Otherwise all identities are deleted and re-created, meaning connectivity is lost.

oblazek replied:
ah yes, got it

squeed committed:
clustermesh: don't create two caches for a single allocator

Before this change, RemoteCache would create and start a second cache on
an existing Allocator, which is a waste of resources.

So, just pass through to the Allocator's existing cache directly.
There's no need to duplicate etcd watches or the in-memory cache.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

tklauser reviewed (approved):
2nd commit LGTM as well, thanks.

aditighag reviewed (approved):
Approving for my code review group. (I'm not sure I fully understand the new changes in `pkg/allocator/allocator.go` in the 2nd commit. Hopefully other code review areas can take a closer look. )

squeed committed:
github/conformance-multicluster: wait for cluster to converge

When two clusters are connected, it causes a full agent rollout. So,
wait for that to finish with "cilium status" before proceeding with
connectivity tests.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

YutaroHayakawa reviewed (approved):
Sorry for the late resposse. Changes looks good to me (including the test changes) thanks!

nbusseneau commented on .github/workflows/conformance-multicluster.yaml:
```diff
@@ -322,6 +322,8 @@ jobs:
 
       - name: Wait for cluster mesh status to be ready
         run: |
+          cilium status --wait --context ${{ steps.contexts.outputs.context1 }}
+          cilium status --wait --context ${{ steps.contexts.outputs.context2 }}
```
While this is fine for now, I suppose a more proper fix would be to have the `clustermesh status --wait` command automatically also imply a `status --wait` directly in CLI code. cc @tklauser for thoughts.

squeed replied:
Well, ultimately, they do different things. Ideally there would be a `--wait` in `clustermesh join`
